docs/zq/language.md

```diff
@@ -1,341 +1,2091 @@
+undefined order), merged (in a defined order) by one or more sort keys,
+or joined using relational-style join logic.
 
-TBD: this is linked from below
+Generally speaking, a [flow graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph)
+defines a directed acyclic graph (DAG) composed
+of data sources and operator nodes.  The Zed syntax leverages "fat arrows",
+i.e., `=>`, to indicate the start of a parallel legs of the data flow.
 
-## Record Mutations
+That said, the Zed language is
+[declarative](https://en.wikipedia.org/wiki/Declarative_programming)
+and the Zed compiler optimizes the data flow computation
+&mdash; e.g., often implementing a Zed program differently than
+the flow implied by the pipeline yet reaching the same result &mdash;
+much as a modern SQL engine optimizes a declarative SQL query.
 
-TBD: this is linked from below
+Zed is also intended to provide a seamless transition from a simple search experience
+(e.g., typed into a search bar or as the query argument of the  `zq` command-line
+tool) to more a complex analytics experience composed of complex joins and aggregations
+where the Zed language source text would typically be authored in a editor and
+managed under source-code control.
 
-## Type Fusion
+Like an email or Web search, a simple keyword search is just the word itself,
+e.g.,
+```
+sample.com
+```
+is a search for the string "sample.com" and
+```
+sample.com urgent
+```
+is a search for values with the both strings "sample.com" and "urgent" present.
 
-TBD: this is linked from below
+Unlike typical log search systems, the Zed language operators are uniform:
+you can specify an operator including keyword search terms, Boolean predicates,
+etc. using the same syntax at any point in the pipeline as
+[described below](#7-search-expressions)
 
-## Record Literal
+For example,
+the predicate `message_length > 100` can simply be tacked onto the keyword search
+from above, e.g.,
+```
+sample.com urgent message_length > 100
+```
+finds all values containing the string "sample.com" and "urgent" somewhere in them
+provided further that the field `message_length` is a numeric value greater than 100.
+A related query that performs an aggregation could be more formally
+written as follows:
+```
+search "sample.com" AND "urgent"
+| where message_length > 100
+| summarize kinds:=union(type) by net:=network_of(srcip)
+```
+which computes an aggregation table of different message types (e.g.,
+from a hypothetical field called `type`) into a new, aggregated field
+called `kinds` and grouped by by the network of all the source IP address
+in the input
+(e.g.,from a hypothetical field called `srcip`) as a derived field called `net`.
 
-TBD: this is linked from below
+The short-hand query from above might be typed into a search box while the
+latter query might be composed in a query editor or in Zed source files
+maintained in Github.  Both forms are valid Zed queries.
```

Comment:
```suggestion
maintained using git ("gitops for data").  Both forms are valid Zed queries.
```

Reply:
I'm not sure Noah would go for this :-)

Reply:
I think either "in GitHub" or "in Git" is fine, though my heart lies with the more generic "in a version control system".